### PR TITLE
Upgrade linters to ESLint with stricter code style

### PR DESCRIPTION
**Pull request by matthewlane**
## Type of change
- Build related changes

## Description of change
- Updates the quality/linting system from ([JSHint](http://jshint.com/) && [JSCS](http://jscs.info/)) to [ESLint](http://eslint.org/)
- Updates linting system to use a stricter code style

## Other information
This PR is currently configured to use [JavaScript Standard Style](https://standardjs.com/) with semicolons. Compare this to other common linting configurations with number of linting problems found against the `0.21.0` release, and the number after running [`eslint --fix src`](http://eslint.org/docs/user-guide/command-line-interface#--fix):

| config | # of problems | # after `--fix` |
| :--- | :--- | :--- |
| [eslint:recommended](http://eslint.org/docs/rules/) | 121 errors, 0 warnings | 121 errors, 0 warnings |
| [Standard](https://www.npmjs.com/package/eslint-config-standard) (with semicolons) | 2254 errors, 0 warnings | 321 errors, 0 warnings |
| [Airbnb](https://www.npmjs.com/package/eslint-config-airbnb) | 7214 errors, 376 warnings | 1897 errors, 257 warnings |

Other configs, such as [Canonical](https://www.npmjs.com/package/eslint-config-canonical) mentioned in #776, or a custom rule set are possible.

Compare rule sets with https://sqren.github.io/eslint-compare/

Requesting feedback from the community and @prebid/core  on which code style Prebid.js should use.

### CODE REVIEW

**Comment by mkendall07** [comment]
Hey @snapwich , @protonate thoughts on this?

**Comment by dbemiller** [comment]
Changes look great. ESLint is a really nice tool.

My only real concern is not having it in the CI process anywhere. If I check out your branch and run `gulp lint`, it prints out the 2254 errors like you mentioned. But... it looks like the CI builds are still passing.

If there's nothing automation preventing these errors, then I don't think people will respect them. It's unrealistic to say "please don't add any _more_ errors," because someone won't notice the difference between 2254 and 2256.

Do you think it's worth fixing these now? Or do you want to run `eslint --fix` locally (to get it down to ~300) and then open another issue about it?

**Comment by mkendall07** [comment]
I agree with @dbemiller on this. Can we commit the  `eslint --fix` changes, add exceptions for the ~300 existing errors and enforce strict checking in travis (add the `lint` gulp task into `run-tests`  which travis invokes). ?

**Comment by protonate** [comment]
@dbemiller looks like Travis CI failed on ES lint errors. Would you update the PR to pass CI and then we can merge. Thanks!

**Comment by dbemiller** [comment]
@protonate Yeah, that failure is on purpose. The weird behavior seems to stem from the parallelized gulp tasks stepping on each other's toes.

I think I straightened them out... but want to re-run the Travis build 4-5 times to make sure it fails reliably. Then I'll fix them and run it 4-5 times to make sure it passes reliably.

**Comment by dbemiller** [comment]
Looks to be good now ^^. Someone might want to sanity check the gulpfile before merging, though, since I declared some task dependencies that don't necessarily make a ton of sense.

I don't think they'll _hurt_ anything (besides performance)... but it'd be good to have another set of eyes on them.

**Comment by dbemiller** [comment]
hey @snapwich, do you think you'll be able to get around to this sometime soon? Nearly every PR which gets merged has style issues in it, and keeping this branch up to date is getting tedious (if I don't fix them, Travis will fail all the PRs until they _are_ fixed) Once this gets into master, CI will take care of it for us.

All the changes to actual JS code were generated by `eslint --fix`. When merging from master, I used `git checkout --theirs` and then ran `eslint --fix` again. Any errors which eslint couldn't fix automatically were dealt with by adding exceptions to the .eslintrc (which can be fixed in later PRs) to keep the changeset as safe as possible.

File: package.json [hunk 1]
```diff
@@ -31,17 +31,21 @@
...
+    "eslint-plugin-import": "^2.2.0",
+    "eslint-plugin-node": "^4.2.2",
+    "eslint-plugin-promise": "^3.5.0",
+    "eslint-plugin-standard": "^3.0.1",
     "faker": "^3.1.0",
     "fs.extra": "^1.3.2",
     "gulp": "^3.8.7",
     "gulp-babel": "^6.1.2",
-    "gulp-clean": "^0.3.1",
+    "gulp-clean": "^0.3.2",
```
Comment (snapwich) [inline package.json]: If we're updating dependencies, the `yarn.lock` file should probably be updated as well.

Comment (dbemiller) [inline package.json]: good catch ^^. Done.

File: gulpfile.js [hunk 2]
```diff
@@ -189,21 +195,11 @@ gulp.task('watch', function () {
...
-gulp.task('hint', function () {
-  return gulp.src('src/**/*.js')
-    .pipe(jshint('.jshintrc'))
-    .pipe(jshint.reporter('jshint-stylish'))
-    .pipe(jshint.reporter('fail'));
-});
-
-gulp.task('jscs', function () {
+gulp.task('lint', () => {
   return gulp.src('src/**/*.js')
```
Comment (dbemiller) [inline gulpfile.js]: Agreed & done.

**Pull request merged**